### PR TITLE
Add 'out' parameter to ReductionKernel, as in PyOpenCL

### PR DESCRIPTION
**Pull request by vincefn**

### CODE REVIEW

File: pycuda/reduction.py [hunk 1]
```diff
@@ -228,6 +228,7 @@ def __call__(self, *args, **kwargs):
             s2_func = kernel_wrapper(s2_func)
 
         stream = kwargs.get("stream")
+        out = kwargs.pop("out", None)
```
Comment (inducer) [inline pycuda/reduction.py:231]: Could you add this to the docs?

File: pycuda/reduction.py [hunk 2]
```diff
@@ -267,7 +268,9 @@ def __call__(self, *args, **kwargs):
                 macroblock_size = block_count*self.block_size
                 seq_count = (sz + macroblock_size - 1) // macroblock_size
 
-            if block_count == 1:
+            if block_count == 1 and out is not None:
+                result = out
```
Comment (inducer) [inline pycuda/reduction.py:272]: You should check that it has the right size and dtype.

Comment (vincefn) [reply pycuda/reduction.py:274]: I added a check for the dtype. Why the size - only one element is used ?

File: pycuda/reduction.py (hunk 1 above)
Comment (vincefn) [reply pycuda/reduction.py:231]: Done - there was no __call__ doc so I added it and expanded the example

File: doc/source/array.rst [hunk 3]
```diff
@@ -1145,6 +1155,19 @@ Here's a usage example::
...
     my_dot_prod = krnl(a, b).get()
 
+Or by specifying the output::
+
+    from pycuda.curandom import rand as curand
+    a = curand((10, 200), dtype=np.float32)
+    red = ReductionKernel(np.float32, neutral=0,
+                               reduce_expr="a+b",
+                               arguments="float *in")
+    a_sum = gpuarray.empty(10, dtype=np.float32)
```
Comment (inducer) [inline doc/source/array.rst:1165]: How come this is created as size-10, when only 1 element is used?

File: pycuda/reduction.py [hunk 4]
```diff
@@ -267,7 +268,11 @@ def __call__(self, *args, **kwargs):
                 macroblock_size = block_count*self.block_size
                 seq_count = (sz + macroblock_size - 1) // macroblock_size
 
-            if block_count == 1:
+            if block_count == 1 and out is not None:
+                assert out.dtype == self.dtype_out, \
```
Comment (inducer) [inline pycuda/reduction.py:272]: Please use `if`, not `assert`

File: pycuda/reduction.py (hunk 2 above)
Comment (inducer) [reply pycuda/reduction.py:274]: At the very least, there should be a check that the array is non-empty (i.e. size > 0). Additionally, I think it helps set expectations: If someone passes a bigger-than-single-entry array, they might expect all entries to get filled. Plus, it's easy to get and pass a single-entry view of an array if that's what's desired.

Comment (vincefn) [reply pycuda/reduction.py:276]: OK, I added the size test

File: doc/source/array.rst (hunk 3 above)
Comment (vincefn) [reply doc/source/array.rst:1165]: The example fills each element of a_sum by calling the reduction kernel 10 times.

File: pycuda/reduction.py (hunk 4 above)
Comment (vincefn) [reply pycuda/reduction.py]: OK, switched to `if` & `raise ValueError`

File: doc/source/array.rst (hunk 3 above)
Comment (inducer) [reply doc/source/array.rst:1165]: OIC, nvm.

**Comment by inducer** [comment]
CI: https://gitlab.tiker.net/inducer/pycuda/-/merge_requests/24

**Comment by inducer** [comment]
Thanks for your contribution! LGTM, it'll go in if CI passes.

**Pull request merged**